lightning/src/ln/outbound_payment.rs

```diff
@@ -418,6 +428,104 @@ impl OutboundPayments {
 		}
 	}
 
+	pub(super) fn check_retry_payments<R: Deref, ES: Deref, NS: Deref, SP, IH, FH, L: Deref>(
+		&self, router: &R, first_hops: FH, inflight_htlcs: IH, entropy_source: &ES, node_signer: &NS,
+		best_block_height: u32, logger: &L, send_payment_along_path: SP,
+	)
+	where
+		R::Target: Router,
+		ES::Target: EntropySource,
+		NS::Target: NodeSigner,
+		SP: Fn(&Vec<RouteHop>, &Option<PaymentParameters>, &PaymentHash, &Option<PaymentSecret>, u64,
+		   u32, PaymentId, &Option<PaymentPreimage>, [u8; 32]) -> Result<(), APIError>,
+		IH: Fn() -> InFlightHtlcs,
+		FH: Fn() -> Vec<ChannelDetails>,
+		L::Target: Logger,
+	{
+		loop {
+			let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+			let mut retry_id_route_params = None;
+			for (pmt_id, pmt) in outbounds.iter_mut() {
+				if pmt.is_retryable() {
+					if let PendingOutboundPayment::Retryable { pending_amt_msat, total_msat, route_params: Some(params), .. } = pmt {
+						if pending_amt_msat < total_msat {
+							retry_id_route_params = Some((*pmt_id, params.clone()));
+						}
+					}
+				}
+				if retry_id_route_params.is_some() { pmt.increment_attempts(); break }
+			}
+			if let Some((payment_id, route_params)) = retry_id_route_params {
+				core::mem::drop(outbounds);
+				if let Err(e) = self.pay_internal(payment_id, route_params, router, first_hops(), inflight_htlcs(), entropy_source, node_signer, best_block_height, &send_payment_along_path) {
+					log_trace!(logger, "Errored retrying payment: {:?}", e);
+				}
+			} else { break }
+		}
+	}
```

Comment:
The outer loop and lock dropping makes this difficult to read. Isn't really clear to me why it is needed. Can't we simply loop through `pending_outbound_payments` once and call `pay_internal` as needed by passing a `PendingOutboundPayment` to it on each iteration?

Guessing it has something to do with locking in `ChannelManager::send_payment_along_path`, but I'm worried that with all the callbacks and locking this is becoming difficult to understand and maintain.

Reply:
We can't hold the `outbounds` lock going into `pay_internal` because then we'd be holding it during pathfinding. I agree it's weird, we'll also have n^2 runtime here relative to the size of `outbounds`. 

Previously I gathered all the retryable outbounds into a vec under one lock, then retried them one by one, which was a bit cleaner IMO at the expense of an allocation. Change was requested here: https://github.com/lightningdevkit/rust-lightning/pull/1916#discussion_r1070425373. Open to other ways to clean this up.

Reply:
I can see why we wouldn't want to hold the lock during path finding, yeah. It doesn't seem the avoid the clones though? But no `Vec` allocation, I suppose.